### PR TITLE
Render custom images in reactions

### PR DESCRIPTION
**Pull request by sumnerevans**
image

Type: enhancement

Implements MSC4027.

Notes: Render custom images in reactions according to MSC4027

Signed-off-by: Sumner Evans <sumner@beeper.com>

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Render custom images in reactions according to MSC4027 ([\#11087](https://github.com/matrix-org/matrix-react-sdk/pull/11087)). Contributed by @sumnerevans.

### CODE REVIEW

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 1]
```diff
@@ -130,9 +131,18 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
                 onMouseOver={this.onMouseOver}
                 onMouseLeave={this.onMouseLeave}
             >
-                <span className="mx_ReactionsRowButton_content" aria-hidden="true">
-                    {content}
-                </span>
+                {content.startsWith("mxc://") ? (
```
Comment (anoadragon453) [inline src/components/views/messages/ReactionsRowButton.tsx:134]: Does matrix-react/js-sdk not have a more generic MXC URI validation function already?

Comment (sumnerevans) [reply src/components/views/messages/ReactionsRowButton.tsx:134]: It seems like this method is used across the codebase: https://github.com/search?q=repo%3Amatrix-org%2Fmatrix-react-sdk+%22.startsWith%28%5C%22mxc%3A%2F%2F%5C%22%29%22&type=code.

I couldn't find a helper function for this.

File: src/components/views/messages/ReactionsRowButtonTooltip.tsx [hunk 2]
```diff
@@ -43,12 +43,16 @@ export default class ReactionsRowButtonTooltip extends React.PureComponent<IProp
         let tooltipLabel: JSX.Element | undefined;
         if (room) {
             const senders: string[] = [];
+            let customReactionName = "";
```
Comment (SimonBrandner) [inline src/components/views/messages/ReactionsRowButtonTooltip.tsx:46]:
```suggestion
            let customReactionName: string | undefined;
```

I think I'd prefer this but it's more of a personal preference

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 3]
```diff
@@ -106,21 +107,48 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
 
         const room = this.context.getRoom(mxEvent.getRoomId());
         let label: string | undefined;
+        let customReactionName = "";
```
Comment (SimonBrandner) [inline src/components/views/messages/ReactionsRowButton.tsx:110]:
```suggestion
        let customReactionName: string | undefined;
```

I think I'd prefer this but it's more of a personal preference

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 4]
```diff
@@ -106,21 +107,48 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
...
+                    <img
+                        className="mx_ReactionsRowButton_content"
+                        alt={customReactionName || content}
+                        src={imageSrc}
+                        width="16"
+                        height="16"
+                    />
+                );
+            }
+        }
```
Comment (SimonBrandner) [inline src/components/views/messages/ReactionsRowButton.tsx:150]: Could this be an if-else statement?

**Review by SimonBrandner — COMMENTED** [review]
Looks good but I think we'd prefer this under a labs flag

File: src/components/views/messages/ReactionsRowButton.tsx (hunk 4 above)
Comment (sumnerevans) [reply src/components/views/messages/ReactionsRowButton.tsx:150]: The reason I did it this way is that there are two necessary conditions for rendering as an image:

1. Does it start with `mxc://`?
2. If so, then does getting the media src using `mediaFromMxc(content).srcHttp` work?

Both of these have to be true in order for it to make sense to render as an image.

Since JavaScript doesn't have any way to declare a variable that can be used within the block within the `if` condition itself, I thought this would be the best option, since otherwise we would need to duplicate the `span` code in the `else` of both the inner and outer `if`s.

If this was Python for example, I think that doing:

```python
if content.startswith("mxc://") and (image_src := media_from_mxc(content).src_http):
    reaction_content = (<img...>)
else:
    reaction_content = (<span...>)
```

Comment (SimonBrandner) [reply src/components/views/messages/ReactionsRowButton.tsx:150]: Ah, yeah, you're right!

**Comment by sumnerevans** [comment]
Where do I even add tests? It's complaining about not being tested, but it already wasn't tested.

**Review by SimonBrandner — APPROVED** [review]
Looks sane codewise!

I am going to check on Monday if this needs a product review

**Comment by SimonBrandner** [comment]
> Where do I even add tests? It's complaining about not being tested, but it already wasn't tested.

Ah, yeah, you might need to write new tests for the code. Also, please run the `yarn i18n` script

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 5]
```diff
@@ -105,22 +107,50 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
...
         const room = this.context.getRoom(mxEvent.getRoomId());
+        const customReactionImagesEnabled = SettingsStore.getValue("feature_render_reaction_images");
         let label: string | undefined;
+        let customReactionName: string | undefined;
         if (room) {
             const senders: string[] = [];
             for (const reactionEvent of reactionEvents) {
                 const member = room.getMember(reactionEvent.getSender()!);
                 senders.push(member?.name || reactionEvent.getSender()!);
+                if (customReactionImagesEnabled && reactionEvent.event.content?.["com.beeper.reaction.shortcode"]) {
```
[the remark this answers is not in the document]
Comment (SimonBrandner) [reply src/components/views/messages/ReactionsRowButton.tsx:118]: Yeah, that would be nice

**Comment by SimonBrandner** [comment]
> Overall, I think we should try to coordinate with custom emotes as an overall feature. It would be confusing to the user if only custom reactions are added to labs.

Since labs is not meant for regular users, I don't think we care too much...

File: src/components/views/messages/ReactionsRowButton.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (sumnerevans) [reply src/components/views/messages/ReactionsRowButton.tsx:118]: It is an unstable value. It will be `shortcode` when it's stable: https://github.com/matrix-org/matrix-spec-proposals/pull/4027/files#diff-c075783679f0bce493f0a817a6f71200aa434dd8ac13fef5265089a8dc433506R126-R127

[the remark this answers is not in the document]
Comment (sumnerevans) [reply src/components/views/messages/ReactionsRowButton.tsx]: Oh, I see. Done.

**Review by SimonBrandner — APPROVED** [review]
Apart from missing tests looks good codewise

I am going to request the product review

**Comment by SimonBrandner** [comment]
Related: https://github.com/matrix-org/matrix-react-sdk/pull/9240

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 6]
```diff
@@ -105,22 +108,56 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
         }
 
         const room = this.context.getRoom(mxEvent.getRoomId());
+        const customReactionImagesEnabled = SettingsStore.getValue("feature_render_reaction_images");
```
Comment (t3chguy) [inline src/components/views/messages/ReactionsRowButton.tsx:111]: Might be good to hoist this to a common parent and share via props/context, each event having to do this can add up to a lot of localStorage accesses which can cause performance issues

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 7]
```diff
@@ -105,22 +108,56 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
...
+        let customReactionName: string | undefined;
         if (room) {
             const senders: string[] = [];
             for (const reactionEvent of reactionEvents) {
                 const member = room.getMember(reactionEvent.getSender()!);
                 senders.push(member?.name || reactionEvent.getSender()!);
+                if (
+                    customReactionImagesEnabled &&
+                    (reactionEvent.event.content?.[REACTION_SHORTCODE_KEY.name] ||
+                        reactionEvent.event.content?.[REACTION_SHORTCODE_KEY.altName])
```
Comment (t3chguy) [inline src/components/views/messages/ReactionsRowButton.tsx:122]:
```suggestion
                    REACTION_SHORTCODE_KEY.findIn(reactionEvent.getContent())
```

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 8]
```diff
@@ -105,22 +108,56 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
...
+                ) {
+                    customReactionName =
+                        reactionEvent.event.content?.[REACTION_SHORTCODE_KEY.name] ||
+                        reactionEvent.event.content?.[REACTION_SHORTCODE_KEY.altName];
```
Comment (t3chguy) [inline src/components/views/messages/ReactionsRowButton.tsx:126]:
```suggestion
                    customReactionName = REACTION_SHORTCODE_KEY.findIn(reactionEvent.getContent());
```

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 9]
```diff
@@ -105,22 +108,56 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
...
+                {content}
+            </span>
+        );
+        if (customReactionImagesEnabled && content.startsWith("mxc://")) {
+            const imageSrc = mediaFromMxc(content).srcHttp;
+            if (imageSrc) {
+                reactionContent = (
+                    <img
+                        className="mx_ReactionsRowButton_content"
+                        alt={customReactionName || content}
```
Comment (t3chguy) [inline src/components/views/messages/ReactionsRowButton.tsx:152]: `content` being an `mxc://` URL is not a helpful alt text

File: src/components/views/messages/ReactionsRowButtonTooltip.tsx [hunk 10]
```diff
@@ -42,13 +44,24 @@ export default class ReactionsRowButtonTooltip extends React.PureComponent<IProp
         const room = this.context.getRoom(mxEvent.getRoomId());
         let tooltipLabel: JSX.Element | undefined;
         if (room) {
+            const customReactionImagesEnabled = SettingsStore.getValue("feature_render_reaction_images");
```
Comment (t3chguy) [inline src/components/views/messages/ReactionsRowButtonTooltip.tsx:47]: Same as in ReactionsRowButton re hoisting for perf

File: src/components/views/messages/ReactionsRowButtonTooltip.tsx [hunk 11]
```diff
@@ -42,13 +44,24 @@ export default class ReactionsRowButtonTooltip extends React.PureComponent<IProp
...
                 senders.push(name);
+                if (
+                    customReactionImagesEnabled &&
+                    (reactionEvent.event.content?.[REACTION_SHORTCODE_KEY.name] ||
+                        reactionEvent.event.content?.[REACTION_SHORTCODE_KEY.altName])
+                ) {
+                    customReactionName =
+                        reactionEvent.event.content?.[REACTION_SHORTCODE_KEY.name] ||
+                        reactionEvent.event.content?.[REACTION_SHORTCODE_KEY.altName];
+                }
```
Comment (t3chguy) [inline src/components/views/messages/ReactionsRowButtonTooltip.tsx:62]: Please use the same accessor pattern I suggested above

**Review by t3chguy — CHANGES_REQUESTED** [review]

**Comment by t3chguy** [comment]
I think this is really promising, but it does require some test coverage realistically

**Comment by t3chguy** [comment]
Related/conflicting https://github.com/matrix-org/matrix-react-sdk/pull/7903

File: src/components/views/messages/ReactionsRowButton.tsx (hunk 9 above)
Comment (sumnerevans) [reply src/components/views/messages/ReactionsRowButton.tsx:146]: what would you recommend the alt text be? Just empty in the case of no custom reaction name?

Comment (t3chguy) [reply src/components/views/messages/ReactionsRowButton.tsx:146]: `_("Custom reaction")` perhaps? its the edge case given the vast majority will have shortcodes, no?

Comment (sumnerevans) [reply src/components/views/messages/ReactionsRowButton.tsx]: sounds good. I guess it will probably be an edge case.

**Comment by sumnerevans** [comment]
@t3chguy @SimonBrandner this now has tests.

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 12]
```diff
@@ -100,27 +105,58 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
...
                     reactionEvents={reactionEvents}
                     visible={this.state.tooltipVisible}
+                    customReactionImagesEnabled={this.props.customReactionImagesEnabled}
                 />
             );
         }
 
         const room = this.context.getRoom(mxEvent.getRoomId());
         let label: string | undefined;
+        let customReactionName: Optional<string>;
```
Comment (SimonBrandner) [inline src/components/views/messages/ReactionsRowButton.tsx:115]: I think this is not our preferred way of doing an optional var - I think `string | undefined` is often preffered

File: src/components/views/messages/ReactionsRowButton.tsx [hunk 13]
```diff
@@ -100,27 +105,58 @@ export default class ReactionsRowButton extends React.PureComponent<IProps, ISta
...
             const senders: string[] = [];
             for (const reactionEvent of reactionEvents) {
                 const member = room.getMember(reactionEvent.getSender()!);
                 senders.push(member?.name || reactionEvent.getSender()!);
+                if (
+                    this.props.customReactionImagesEnabled &&
+                    REACTION_SHORTCODE_KEY.findIn(reactionEvent.getContent())
+                ) {
+                    customReactionName = REACTION_SHORTCODE_KEY.findIn(reactionEvent.getContent());
+                }
```
Comment (SimonBrandner) [inline src/components/views/messages/ReactionsRowButton.tsx:126]: How about something like `customReactionName = this.props.customReactionImagesEnabled ? REACTION_SHORTCODE_KEY.findIn(....) : undefined`

File: test/components/views/messages/ReactionsRowButton-test.tsx [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        const event = new MouseEvent("mouseover", {
+            bubbles: true,
+            cancelable: true,
+        });
+        reactionButton.dispatchEvent(event);
+
+        expect(root).toMatchSnapshot();
+    });
+
+    it("reaction row button renders custom image reactions correctly", () => {
```
Comment (SimonBrandner) [inline test/components/views/messages/ReactionsRowButton-test.tsx:90]:
```suggestion
    it("renders reaction row button custom image reactions correctly", () => {
```

File: src/components/views/messages/ReactionsRowButton.tsx (hunk 12 above)
Comment (sumnerevans) [reply src/components/views/messages/ReactionsRowButton.tsx:115]: The reason I did this is because the `findIn` function returns `Optional<string>`

File: test/components/views/messages/ReactionsRowButton-test.tsx [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+        // Try hover and make sure that the ReactionsRowButtonTooltip works
+        const reactionButton = root.getByRole("button");
+        const event = new MouseEvent("mouseover", {
+            bubbles: true,
+            cancelable: true,
+        });
+        reactionButton.dispatchEvent(event);
+
+        expect(root).toMatchSnapshot();
```
Comment (t3chguy) [inline test/components/views/messages/ReactionsRowButton-test.tsx:88]:
```suggestion
        expect(root.asFragment()).toMatchSnapshot();
```

File: test/components/views/messages/ReactionsRowButton-test.tsx [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+                key: "👍",
+                rel_type: "m.annotation",
+            },
+        });
+        const root = render(
+            <MatrixClientContext.Provider value={mockClient}>
+                <ReactionsRowButton {...props} />
+            </MatrixClientContext.Provider>,
+        );
+        expect(root).toMatchSnapshot();
```
Comment (t3chguy) [inline test/components/views/messages/ReactionsRowButton-test.tsx:78]:
```suggestion
        expect(root.asFragment()).toMatchSnapshot();
```

**Review by t3chguy — CHANGES_REQUESTED** [review]
And update snapshots please :)

**Review by t3chguy — APPROVED** [review]
This looks sane to me!

**Comment by sumnerevans** [comment]
This PR does not compete with https://github.com/matrix-org/matrix-react-sdk/pull/7903 because that PR doesn't implement [MSC4027](https://github.com/matrix-org/matrix-spec-proposals/pull/4027). For example, that PR does not support shortcodes, has no tests, and is not behind a labs flag.

**Review by daniellekirkwood — APPROVED** [review]

**Comment by daniellekirkwood** [comment]
Thank you for y'alls time and patience on this one. 

Approving for Product 👍
